File: Teacher.java
/**
 * @author Anna Quinlan, MAHS '19
 */

import java.util.Scanner;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
public class Teacher
{
    private ArrayList<Student> students;
    private String firstName, lastName;
    public Teacher (String first, String last){
        students = new ArrayList<Student>();
        firstName = first;
        lastName = last;
    }
    
    public void addStudent (String fN, String lN, boolean grad, boolean music){
        students.add(new Student(fN, lN, grad, music));
    }
    
    public void addInstantiatedStudent (Student stud){
        students.add(stud);
    }
    
    public void removeStudent (int index){
        students.remove(index);
    }
    
    public void addTeacher (){
        students.add(0, new Student(firstName, lastName));
    }
    
    public String getFirstName(){
        return firstName;
    }
    
    public String getLastName(){
        return lastName;
    }
    
    public int getNumberOfStudents(){
        return students.size();
    }
    
    public Student getStudent(int position){
        return students.get(position);
    }
    
    public void sortByLastName (){
        
        Collections.sort(students, new Comparator<Student>() {
                @Override
                public int compare(Student t1, Student t2) {
                    return t1.getLastName().compareTo(t2.getLastName());
            }
        });
    }
   
    public boolean containsSpeaker(){
        for (int i=0; i<students.size(); i++){
            if (students.get(i).isSpeaker()){return true;}
        }
        return false;
    }
    
    public boolean containsMusician(){
        for (int i=0; i<students.size(); i++){
            if (students.get(i).isMusician()){return true;}
        }
        return false;
    }
    
    public boolean isPrincipal(){
        if (lastName.equals("Rick-Kennel")){return true;}
        if (lastName.equals("1 Rick-Kennel")){return true;}
        return false;
    }
    
    public String toStringWithFlags (){
        String output="";
        for (int i=0; i<students.size()-1; i++){
            output+= students.get(i).toStringWithFlags()+", ";
        }
        output+=students.get(students.size()-1);
        return output+"\n";
    }
    
    public String toString (){
        String output="";
        if (isPrincipal()){
            output+="!Principal ";
        }
        for (int i=0; i<students.size()-1; i++){
            output+= students.get(i)+", ";
        }
        output+=students.get(students.size()-1);
        return output+"\n";
    }
}

File: python-library/teacher.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sat Dec  7 08:30:16 2019

@author: annaquinlan
"""
from student import Student

class Teacher:

    def __init__(self, first_name, last_name):
        self.first_name = first_name
        self.last_name = last_name
        self.students = []
        self.has_student_with_priority = False

    def __lt__(self, other):
        if self.has_student_with_priority == other.has_student_with_priority:
            return self.last_name < other.last_name
        return self.has_student_with_priority

    def __str__(self):
        output = ""
        if self.has_student_with_priority:
            output += "!HasPriority "

        for student in self.students[:-1]:
            output += str(student) + ", "

        output += str(self.students[-1])
        return output

    def add_student(self, first, last, has_priority):
        if has_priority:
            self.has_student_with_priority = True

        self.students.append(Student(first, last, has_priority))

    def add_instantiated_student(self, student):
        if student.has_priority:
            self.has_student_with_priority = True

        self.students.append(student)

    def remove_student(self, index):
        if 0 <= index < len(self.students):
            del self.students[index]

    def add_teacher(self):
        self.students.insert(0, Student(self.first_name, self.last_name, False))

    def sort_students(self):
        self.students.sort()

    def get_number_of_students(self):
        return len(self.students)

    def is_principal(self, principal_first, principal_last):
        return (self.first_name.lower() == principal_first.lower() 
                and self.last_name.lower() == principal_last.lower()
                )
